Source/Utils/IntervalsParser.cpp

```diff
@@ -116,3 +122,94 @@ int IntervalsParser::localPeriod (const int n) const
+    for (const auto& instr_element : instr_vec) inconcatenated +=instr_element;
+
+    auto const insplit = WarpXUtilStr::split<std::vector<std::string>>(inconcatenated, std::string(1,m_separator));
+
+    // parse the Intervals string into Slices and store each slice in m_slices,
+    // in order of increasing Slice start value
+    for(const auto& inslc : insplit)
+    {
+        bool isBTD = true;
+        SliceParser temp_slice(inslc, isBTD);
+        if (m_slices.size() > 0)
+        {
+            // find the last index i_slice where
+            // the start value of m_slices[i_slice] is greater than temp_slices' start_value
+            int i_slice = 0;
+            while (temp_slice.getStart() > m_slices[i_slice].getStart() && i_slice < static_cast<int>(m_slices.size()))
+            {
+                i_slice++;
+            }
+            m_slices.insert(m_slices.begin() + i_slice, temp_slice);
+        }
+        else
+        {
+            m_slices.push_back(temp_slice);
+        }
+    }
+    // from the vector of slices, m_slices,
+    // create a vector of integers, m_btd_iterations, containing
+    // the iteration of every back-transformed snapshot that will be saved
+    // the iteration values in m_btd_iterations are
+    // 1. saved in increasing order
+    // 2. unique, i.e. no duplicate iterations are saved
+    for (const auto& temp_slice : m_slices)
+    {
+        const int start = temp_slice.getStart();
+        const int period = temp_slice.getPeriod();
+        int btd_iter_ind;
+        // for Slice temp_slice in m_slices,
+        // determine the index in m_btd_iterations where temp_slice's starting value goes
+        //
+        // Implementation note:
+        // assuming the user mostly lists slices in ascending order,
+        // start at the end of m_btd_iterations and search backward
+        if (m_btd_iterations.size() == 0)
+        {
+            btd_iter_ind = 0;
+        }
+        else
+        {
+            btd_iter_ind = m_btd_iterations.size() - 1;
+            while (start < m_btd_iterations[btd_iter_ind] and btd_iter_ind>0)
+            {
+                btd_iter_ind--;
+            }
+        }
+        // insert each iteration contained in temp_slice into m_btd_iterations
+        // adding them in increasing sorted order and not adding any iterations
+        // already contained in m_btd_iterations
+        for (int ii = start; ii <= temp_slice.getStop(); ii += period)
+        {
+            if (m_btd_iterations.size() > 0)
+            {
+                // find where iteration ii should go in m_btd_iterations
+                while (ii > m_btd_iterations[btd_iter_ind] && btd_iter_ind < static_cast<int>(m_btd_iterations.size()))
+                {
+                    btd_iter_ind++;
+                }
+                if (ii != m_btd_iterations[btd_iter_ind])
+                {
+                    m_btd_iterations.insert(m_btd_iterations.begin() + btd_iter_ind, ii);
+                }
```

Comment:
Looking at this right now with @WeiqunZhang @RevathiJambunathan :
- `m_btd_iterations[...]` should be `m_btd_iterations.at(...)` because it can catch at runtime out-of-bounds access, and the loop here is cheap
- there are out-of-bounds accesses to `m_btd_iterations` in both the `while` condition and the ìf` later on

Let's find a safer formulation.

Reply:
There are two issues. `btd_iter_ind < static_cast<int>(m_btd_iterations.size())` should be checked before `m_btd_iterations[btd_iter_ind]` is used. The second issue is `btd_iter_ind` could be as big as the size of the vector because of `++` in the loop body. So the next if test `if (ii != m_btd_iterations[btd_iter_ind])` is not safe.

Reply:
Fix coming in #4216